contracts/adapters/SwapAdapter.sol

```diff
@@ -0,0 +1,388 @@
+        address feeHandlerRouter;
+        uint256 amountOut;
+        uint256 swapAmount;
+        bytes path;
+        IV3SwapRouter.ExactInputParams params;
+        bytes depositData;
+        address ERC20HandlerAddress;
+        uint256 leftover;
+    }
+
+    error CallerNotAdmin();
+    error AlreadySet();
+    error TokenInvalid();
+    error PathInvalid();
+    error MsgValueLowerThanFee(uint256 value);
+    error InsufficientAmount(uint256 amount);
+    error FailedFundsTransfer();
+    error AmountLowerThanFee(uint256 amount);
+
+    event TokenResourceIDSet(address token, bytes32 resourceID);
+    event TokensSwapped(address token, uint256 amountOut);
+
+    constructor(
+        IBridge bridge,
+        address weth,
+        IV3SwapRouter swapRouter,
+        INativeTokenAdapter nativeTokenAdapter
+    ) {
+        _bridge = bridge;
+        _weth = weth;
+        _swapRouter = swapRouter;
+        _nativeTokenAdapter = nativeTokenAdapter;
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    modifier onlyAdmin() {
+        if (!hasRole(DEFAULT_ADMIN_ROLE, msg.sender)) revert CallerNotAdmin();
+        _;
+    }
+
+    // Admin functions
+    function setTokenResourceID(address token, bytes32 resourceID) external onlyAdmin {
+        if (tokenToResourceID[token] == resourceID) revert AlreadySet();
+        tokenToResourceID[token] = resourceID;
+        emit TokenResourceIDSet(token, resourceID);
+    }
+
+    /**
+        @notice Function for depositing tokens, performing swap to ETH and bridging the ETH.
+        @param destinationDomainID  ID of chain deposit will be bridged to.
+        @param recipient Recipient of the deposit.
+        @param  token Input token to be swapped.
+        @param  tokenAmount Amount of tokens to be swapped.
+        @param amountOutMinimum Minimal amount of ETH to be accepted as a swap output.
+        @param pathTokens Addresses of the tokens for Uniswap swap. WETH address is used for ETH.
+        @param pathFees Fees for Uniswap pools.
+    */
+    function depositTokensToEth(
+        uint8 destinationDomainID,
+        address recipient,
+        address token,
+        uint256 tokenAmount,
+        uint256 amountOutMinimum,
+        address[] calldata pathTokens,
+        uint24[] calldata pathFees
+    ) external {
+        if (tokenToResourceID[token] == bytes32(0)) revert TokenInvalid();
+
+        // Swap all tokens to ETH (exact input)
+        IERC20(token).safeTransferFrom(msg.sender, address(this), tokenAmount);
+        IERC20(token).safeApprove(address(_swapRouter), tokenAmount);
+
+        uint256 amount = swapTokens(
+            pathTokens,
+            pathFees,
+            token,
+            _weth,
+            tokenAmount,
+            amountOutMinimum,
+            0
+        );
+
+        IWETH(_weth).withdraw(amount);
+
+        // Make Native Token deposit
+        _nativeTokenAdapter.depositToEVM{value: amount}(destinationDomainID, recipient);
```

Comment:
@mpetrunic @viatrix 
For me use case here is always I want to send X amount of ETH to some other chain (same with message).
So with this in mind, does it make sense to have logic to send only `X` defined on function call and return swap leftover back to user.

I guess this could led to situations where swap doesn't get enough tokens out of it, but then again it would also probably fail on execution as your amount is off.

Reply:
Makes sense. For this use case it's also possible to swap using exactOutput. The user could specify max amount of input to be swapped and desired exact amount to send. We swap amount+sygma fee and send leftover tokens to the user. The difference is that in your case the user would get eth, in exactOutput case the user would get tokens.
Also, do we need both use cases? (1 - swap and send as much as possible, 2 - send exact amount)?

Reply:
@mak iant sygma contract on destination returning excess eth amount to the recipient on destination?

Right now, only with usdc on destination, excess funds would end up on default message receiver.

I think it might make sense to retain excess amount on source for us or user and send to destination exact amount user requested

Reply:
No problem, will do if you agree on that.
1. Do we need to support also a use case for "swap all and send all" (how it's built now)? or just "send exact amount" case?
2. What is better to return to the user: original currency (swap only what's necessary for exact amount, +fee if swap for eth) or swapped currency (swap all, return excess amount)?

Reply:
Hey @viatrix has short sync with @mpetrunic, so:
1. There is no need for _swap all and send all_ functionality. You are always intentionally sending X amount for some action to another chain. Sorry for not being clear enough from the beginning on this, but we are also designing this feature on the fly :see_no_evil: 
2. It is better to return the original currency. That way, we are not forcing users to end up with tokens that they didn't have in the first place, which is definitely a better UX.

Reply:
Sounds good, will do! Yes, I also thought that it's better to return original currency, I'll do it this way.